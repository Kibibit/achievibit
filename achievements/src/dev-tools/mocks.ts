export class Shall {
  grantedAchievements: { [username: string]: any };

  grant(username, achievementObject) {
    this.grantedAchievements = this.grantedAchievements || {};
    this.grantedAchievements[username] = achievementObject;
  }
}

export class PullRequest {
  title = 'this is a happy little title';
  id = 'test';
  number: number;
  url = 'url';
  organization: { username: string };
  description = '';
  creator = {
    username: 'creator'
  };
  reviewers = [ {
    'username': 'reviewer'
  } ];
  merged: any;
  reviews: any;
  comments: any[];
  inlineComments: any[];
  reactions: any[];
  commits: any[];
  labels: string[];
  createdOn: Date;
  files: { name: string }[];
}
