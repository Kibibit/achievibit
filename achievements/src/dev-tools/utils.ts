export function createComment(username: string, message: string) {
  return {
    'author': {
      'username': username
    },
    'message': message
  };
}