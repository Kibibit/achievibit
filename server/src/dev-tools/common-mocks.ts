import { ArgumentsHost, Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { prop as PersistInDb, ReturnModelType } from '@typegoose/typegoose';
import { Exclude, Expose } from 'class-transformer';

import { BaseModel, BaseService } from '@kb-abstracts';

export const mockResponse = {
  status: jest.fn().mockReturnThis(),
  json: jest.fn().mockReturnThis(),
  sendFile: jest.fn().mockReturnThis(),
  mockClear: () => {
    mockResponse.status.mockClear();
    mockResponse.json.mockClear();
    mockResponse.sendFile.mockClear();
  }
};

export const hostMock = (req, res, roles?: string[]): ArgumentsHost => {
  const ctx: any = {}
  ctx.switchToHttp = jest.fn().mockReturnValue({
      getRequest: () => req,
      getResponse: () => res
  });
  ctx.getHandler = jest.fn().mockReturnValue({ roles }) as Function;

  return ctx;
};

@Exclude()
export class MockModel extends BaseModel {
  @PersistInDb()
  mockPrivateAttribute: string;

  @Expose()
  @PersistInDb({ required: true })
  mockAttribute: string;

  @Exclude()
  @PersistInDb()
  updatedDate?: Date;

  constructor(partial: Partial<MockModel> = {}) {
    super();
    Object.assign(this, partial);
  }
}

@Injectable()
export class MockService extends BaseService<MockModel> {
  constructor(
    @InjectModel(MockModel.modelName)
    private readonly injectedModel: ReturnModelType<typeof MockModel>
  ) {
    super(injectedModel, MockModel);
  }
}
