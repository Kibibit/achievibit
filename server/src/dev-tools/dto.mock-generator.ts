import { Chance } from 'chance';
import { kebabCase, times } from 'lodash';
import { ObjectId } from 'mongodb';
import RandExp from 'randexp';

import { PullRequest, Repo, User } from '@kb-models';

type IAchievementFiles = Record<string, Record<string, unknown>>;

interface IChanceMixin {
  mongoObjectId: () => ObjectId;
  mongodbUrl: () => string;
  user: () => User;
  users: () => User[];
  repo: () => Repo;
  repos: () => Repo[];
  pullRequest: () => PullRequest;
  achievementScripts: (numOfAchievements?: number) => IAchievementFiles;
  randexp: (regex: RegExp) => RandExp;
}

const chance = new Chance() as Chance.Chance & IChanceMixin;

export const DtoMockGenerator: Chance.Chance & IChanceMixin =
  chance as Chance.Chance & IChanceMixin;

const customMixin: IChanceMixin & Chance.MixinDescriptor = {
  mongodbUrl(): string {
    return chance.url({ protocol: 'mongodb' + (chance.bool() ? '+srv' : '') });
  },
  randexp(regex: RegExp): RandExp {
    const randExp = new RandExp(regex);
    randExp.max = 10;

    return randExp;
  },
  mongoObjectId(): ObjectId {
    return new ObjectId(chance.string({
      pool: 'abcdefABCDEF123456789',
      length: 24
    }));
  },
  user(): User {
    const isOrg = chance.bool();

    return new User({
      _id: chance.mongoObjectId(),
      username: chance.name(),
      avatar: chance.url(),
      organization: isOrg,
      repos: times(chance.integer({ min: 0, max: 10 }), () => chance.animal()),
      url: chance.url(),
      users: isOrg ?
        times(chance.integer({ min: 0, max: 10 }), () => chance.animal()) :
        undefined,
      achievements: [],
      token: chance.string()
    });
  },
  users(): User[] {
    return times(chance.integer({ min: 0, max: 10 }), () => chance.user());
  },
  repo(): Repo {
    return new Repo({
      _id: chance.mongoObjectId(),
      name: chance.string(),
      fullname: chance.string(),
      organization: chance.company(),
      url: chance.url()
    });
  },
  repos(): Repo[] {
    return times(chance.integer({ min: 0, max: 10 }), () => chance.repo());
  },
  pullRequest() {
    return new PullRequest({
      createdOn: chance.date(),
      creator: chance.name(),
      description: chance.paragraph(),
      number: chance.integer(),
      prid: chance.guid(),
      url: chance.url()
    });
  },
  achievementScripts(
    numOfAchievements: number = chance.integer({ min: 1, max: 25 })
  ): IAchievementFiles {
    const scriptsObject = {};

    times(
      numOfAchievements,
      () => scriptsObject[ `${ kebabCase(chance.animal()) }.achievement` ] = {}
    );

    return scriptsObject;
  }
};

chance.mixin(customMixin);
