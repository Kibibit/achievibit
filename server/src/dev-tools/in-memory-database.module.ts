import { MongooseModule, MongooseModuleOptions } from '@nestjs/mongoose';
import { MongoMemoryServer } from 'mongodb-memory-server';
import mongoose from 'mongoose';

let mongod: MongoMemoryServer;

export const createInMemoryDatabaseModule =
  (options: MongooseModuleOptions = {}) => MongooseModule.forRootAsync({
    useFactory: async () => {
      mongod = new MongoMemoryServer();
      const mongoUri = await mongod.getUri();
      return {
        uri: mongoUri,
        ...options
      }
    }
  });

export const closeInMemoryDatabaseConnection = async () => {
  await mongoose.disconnect();
  if (mongod) await mongod.stop();
}
