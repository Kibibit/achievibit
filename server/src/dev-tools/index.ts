/**
 * @file Automatically generated by barrelsby.
 */

export * from './common-mocks';
export * from './dto.mock-generator';
export * from './in-memory-database.module';
export * from './captured-events/pull-request-assignee-added.event';
export * from './captured-events/pull-request-assignee-removed.event';
export * from './captured-events/pull-request-created-organization.event';
export * from './captured-events/pull-request-created.event';
export * from './captured-events/pull-request-edited.event';
export * from './captured-events/pull-request-label-added.event';
export * from './captured-events/pull-request-label-removed.event';
export * from './captured-events/pull-request-labels-initialized.event';
export * from './captured-events/pull-request-merged.event';
export * from './captured-events/pull-request-review-request-added.event';
export * from './captured-events/pull-request-review-request-removed.event';
export * from './captured-events/review-comment-added.event';
export * from './captured-events/review-comment-edited.event';
export * from './captured-events/review-comment-removed.event';
export * from './captured-events/review-submitted.event';
export * from './captured-events/wehbook-ping.event';
