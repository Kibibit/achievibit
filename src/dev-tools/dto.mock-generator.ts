import { Chance } from 'chance';
import { kebabCase, times } from 'lodash';
import { ObjectId } from 'mongodb';
import RandExp from 'randexp';

import { RepoDto, User } from '@kb-models';

interface IChanceMixin {
  mongoObjectId: () => ObjectId;
  mongodbUrl: () => string;
  user: () => User;
  users: () => User[];
  repoDto: () => RepoDto;
  repoDtos: () => RepoDto[];
  achievementScripts: (numOfAchievements?: number) => { [key: string]: {} };
  randexp: (regex: RegExp) => RandExp;
}

const chance = new Chance() as Chance.Chance & IChanceMixin;

export const DtoMockGenerator: Chance.Chance & IChanceMixin = chance as Chance.Chance & IChanceMixin;

const customMixin: IChanceMixin & Chance.MixinDescriptor = {
  mongodbUrl(): string {
    return chance.url({ protocol: 'mongodb' + (chance.bool() ? '+srv' : '') });
  },
  randexp(regex: RegExp): RandExp {
    const randExp = new RandExp(regex);
    randExp.max = 10;

    return randExp;
  },
  mongoObjectId(): ObjectId {
    return new ObjectId(chance.string({ pool: 'abcdefABCDEF123456789', length: 24 }));
  },
  user(): User {
    const isOrg = chance.bool();

    return new User({
      _id: chance.mongoObjectId(),
      __v: '0',
      username: chance.name(),
      avatar: chance.url(),
      organization: isOrg,
      repos: times(chance.integer({ min: 0, max: 10 }), () => chance.animal()),
      url: chance.url(),
      users: isOrg ?
        times(chance.integer({ min: 0, max: 10 }), () => chance.animal()) :
        undefined,
      achievements: [],
      token: chance.string()
    });
  },
  users(): User[] {
    return times(chance.integer({ min: 0, max: 10 }), () => chance.user());
  },
  repoDto(): RepoDto {
    return new RepoDto({
      _id: chance.mongoObjectId(),
      __v: 0,
      name: chance.string(),
      fullname: chance.string(),
      organization: chance.company(),
      url: chance.url()
    });
  },
  repoDtos(): RepoDto[] {
    return times(chance.integer({ min: 0, max: 10 }), () => chance.repoDto());
  },
  achievementScripts(numOfAchievements: number = chance.integer({ min: 1, max: 25 })): { [key: string]: {} } {
    const scriptsObject = {};

    times(numOfAchievements, () => scriptsObject[`${kebabCase(chance.animal())}.achievement`] = {});

    return scriptsObject;
  }
};

chance.mixin(customMixin);
