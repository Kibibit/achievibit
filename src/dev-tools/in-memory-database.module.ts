import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { MongoMemoryServer } from 'mongodb-memory-server';
import mongoose from 'mongoose';

let mongod: MongoMemoryServer;
let uri: string;

async function testDBFactory() {
  mongod = new MongoMemoryServer();
  uri = await mongod.getConnectionString();
  await mongoose.connect(uri, {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });
  return { uri };
}

@Module({
  imports: [
    MongooseModule.forRootAsync({
      useFactory: testDBFactory
    })
  ]
})
export class InMemoryDatabaseModule {
  static async closeDatabase() {
    await mongoose.connection.dropDatabase();
    await mongoose.connection.close();
    await mongod.stop();
    mongod = null;
    uri = null;
  }
}
