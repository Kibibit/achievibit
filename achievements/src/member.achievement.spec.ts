import moment from 'moment';

import { PullRequest, Shall } from './dev-tools/mocks';
import { member } from './member.achievement';

describe('member achievement', () => {
  it('should not be granted if PR opened less than 2 weeks ago', () => {
    const testShall = new Shall();
    const pullRequest = new PullRequest();

    pullRequest.createdOn = moment().subtract(13, 'days').toDate();

    member.check(pullRequest, testShall);
    expect(testShall.grantedAchievements).toBeUndefined();
  });

  it('should be granted if PR opened more than 2 weeks ago', () => {
    const testShall = new Shall();
    const pullRequest = new PullRequest();

    pullRequest.createdOn = moment().subtract(15, 'days').toDate();

    member.check(pullRequest, testShall);
    expect(testShall.grantedAchievements.creator).toMatchSnapshot();
  });
});
