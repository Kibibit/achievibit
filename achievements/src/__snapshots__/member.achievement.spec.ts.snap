// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`member achievement should be granted if PR opened more than 2 weeks ago 1`] = `
Object {
  "avatar": "images/achievements/member.achievement.jpg",
  "description": "A pull request you've created 2 weeks ago is finally merged",
  "name": "Member pull request #undefined?",
  "relatedPullRequest": "test",
  "short": "Member Commits? member Push? member PR? ohh I member",
}
`;
